models/GroupsService.php

```diff
@@ -74,48 +95,47 @@ public function deleteGroup(\core_kernel_classes_Resource $group)
      *
      * @access public
      * @author Joel Bout, <joel.bout@tudor.lu>
-     * @param  Class clazz
+     * @param  core_kernel_classes_Class $clazz
      * @return boolean
      */
     public function isGroupClass(core_kernel_classes_Class $clazz)
     {
-        return $clazz->equals($this->getRootClass()) || $clazz->isSubClassOf($this->getRootClass());
+        return $clazz->equals($this->getRootClass())
+            || $clazz->isSubClassOf($this->getRootClass());
     }
 
     /**
-     * get the groups of a user
+     * Get the groups of a user
      *
      * @access public
      * @author Joel Bout, <joel.bout@tudor.lu>
-     * @param  string userUri
+     * @param  User $user
      * @return array resources of group
      */
     public function getGroups(User $user)
     {
-        $groups = $user->getPropertyValues(self::PROPERTY_MEMBERS_URI);
-        array_walk($groups, function (&$group) {
-            $group = new core_kernel_classes_Resource($group);
-        });
-        return $groups;
+        return array_map(
+            function (string $group): core_kernel_classes_Resource {
+                return $this->getModel()->getResource($group);
+            },
+            $user->getPropertyValues(self::PROPERTY_MEMBERS_URI)
+        );
     }
-    
+
     /**
-     * gets the users of a group
+     * Gets the users of a group
      *
      * @param string $groupUri
-     * @return array resources of users
+     * @return core_kernel_classes_Resource[] resources of users
      */
-    public function getUsers($groupUri)
+    public function getUsers(string $groupUri): array
     {
-        $subjectClass = TestTakerService::singleton()->getRootClass();
```

Comment:
I was mostly doing that to help with testing (so it is easier to inject it from tests).

However, even if I'd like to, I'm not completely decoupling this (by removing the call to `TestTakerService::singleton()` in the constructor or adding this to the DI container) because there are other places (both in this extension and in others) that create instances of this service via `GroupsService::singleton()`.

Reply:
For testing you can use `ServiceManagerMockTrait` and I would recommend to use the DI container instead of creating a constructor that is only used by tests. This is not a standard way we generally work for this.

Reply:
Rolled back changes to use constructor injection in cb5be793d4c39badf9e7fb89109b0f77246b9077.

Reply:
Marking this thread as solved, please do reopen it if you think there's something missing related to this topic.